Resolve configured directories relative to a specific location

benbc opened the pull request:
Previously we resolved relative to the working directory which made it
hard to manipulate this in tests and may have been unexpected for
embedded users. Now we use a specific location, provided by the internal
unsupported.dbms.directories.neo4j_home setting.

neo4j_home is set to the working directory for server, but the database
directory when running embedded or in the desktop application.

benbc committed:
Resolve configured directories relative to a specific location

Previously we resolved relative to the working directory which made it
hard to manipulate this in tests and may have been unexpected for
embedded users. Now we use a specific location, provided by the internal
unsupported.dbms.directories.neo4j_home setting.

neo4j_home is set to the working directory for server, but the database
directory when running embedded or in the desktop application.

benbc commented:
@martinfurmanski @tinwelint @zhenlineo @Mats-SX @davidegrohmann This is my proposed fix to the directory madness. I'd appreciate your feedback, please. Detailed review not needed yet as there are more commits coming with subsequent clean-up.

Mats-SX commented:
@benbc Looks good to me. Did you also consider HA and Core-edge setups? I'm not positive if they would be covered by the changes made to `PlatformModule` and `ConfigLoader`.

benbc commented:
@Mats-SX I am under the impression that every time we create a kernel we go through that `PlatformModule` constructor, but I could well be wrong. Can you point me in the code at somewhere to start looking?

Mats-SX commented:
@benbc You are probably correct. I scanned through some HA-factories, and they do indeed go through `PlatformModule`. I'm sure some of the other people you tagged could confirm.

benbc committed:
Provide default value for dbms.directories.import in neo4j.conf

We can't give a default for the import dir because of the semantics of
this setting. The setting is overloaded to turn import constraint on and
off as well as configuring which directory it's constrained to. If we
provide a default then there is no way for a user to turn the behaviour
off (if they explicitly configure it to null, say, they will just get
the default value).

So we can't provide a default at the point of definition and we can't
provide the default dynamically in the server either because that would
also leave the user with no way to turn the feature off. However it's
reasonable in the server to default this to an import directory, so
we'll do that by modifying the default neo4j.conf.

benbc committed:
Remove some config overrides that are no longer necessary

We now provide identical values to these in the defaults.

benbc committed:
Unconditionally set the auth store location in server

This is an internal setting which should never be set externally. In
server it will always be at data/dbms/auth.

benbc committed:
Remove explicit directory configuration in tests that is no longer necessary

We have modified the kernel so that it will put other directories under
the database directory when run embedded. So the tests will now
automatically clean up the contents of those directories in their
default locations.

tinwelint commented:
@benbc love it. `debug.log` works splendid. I tried an inconsistencies report, although it still ended up in the wrong place. Can you have a look at that too? See somewhere around https://github.com/neo4j/neo4j/blob/3.1/community/consistency-check/src/main/java/org/neo4j/consistency/report/InconsistencyLogger.java

benbc commented:
@tinwelint Thanks for the pointer, I'll take a look.

benbc committed:
Stop tests from leaving log directories lying around

benbc commented:
@srbaker Please can you review this.
